core/src/main/scala/kafka/server/AbstractFetcherThread.scala

```diff
@@ -132,7 +132,7 @@ abstract class AbstractFetcherThread(name: String,
                       case None => currentPartitionFetchState.offset
                     }
                     partitionMap.put(topicAndPartition, new PartitionFetchState(newOffset))
-                    fetcherLagStats.getFetcherLagStats(topic, partitionId).lag = partitionData.highWatermark - newOffset
+                    fetcherLagStats.getFetcherLagStats(topic, partitionId).lag = Math.max(0, partitionData.highWatermark - newOffset)
```

Comment:
@guozhangwang Thanks for your review. Problem fixed.

I tried to understand the how HW and messageOffset is updated but likely have missed understood the code. Thanks much for your clarification.